File: node/mods/rustscript2/lib/opcodes/checksender.js

module.exports = {

  name: "CHECKSENDER",
  description: "Check transaction sender matches supplied publickey.",

  exampleScript: {
    op: "CHECKSENDER",
    publickey: "<publickey>"
  },

  exampleRequired: {},

  schema: {
    script: {
      publickey: "string"
    },
    required: {}
  },

  execute: function (node, context) {
    const sender =
      (context && context.tx && context.tx.sender) ||
      (context && context["tx.sender"]) ||
      (context && context.sender) ||
      null;

    const required = node.publickey || null;
    if (!required) return false;

    if (!sender) return false;
    return String(sender).toLowerCase() === String(required).toLowerCase();
  }
};

File: rust/saito-core/src/core/consensus/scripting/opcodes/checksender.rs
use crate::core::consensus::block::Block;
use crate::core::consensus::transaction::Transaction;
use crate::core::defs::PrintForLog;
use serde_json::Value;

pub struct CheckSender {
    pub name: String,
    pub description: String,
    pub script: String,
    pub schema: Value,
}

impl CheckSender {
    pub fn validate(context: &Value, tx: Option<&Transaction>, _blk: Option<&Block>) -> u8 {
        let publickey = context["script"]["publickey"].as_str().unwrap_or("");
        if publickey.is_empty() {
            return 0;
        }

        let Some(tx) = tx else {
            return 0;
        };

        let required = publickey.to_ascii_lowercase();

        for slip in &tx.from {
            if slip.public_key.to_base58().to_ascii_lowercase() == required {
                return 1;
            }
        }

        0
    }
}
